lib/DataFileUtil/DataFileUtilImpl.py

```diff
@@ -245,6 +251,271 @@ def make_handle(self, shock_data, token):
     def make_ref(self, object_info):
         return str(object_info[6]) + '/' + str(object_info[0]) + \
             '/' + str(object_info[4])
+
+    def _get_staging_file_path(self, token_user, staging_file_subdir_path):
+        """
+        _get_staging_file_path: return staging area file path
+
+        directory pattern: /data/bulk/user_name/file_name
+
+        """
+        return self.STAGING_FILE_PREFIX + token_user + '/' + staging_file_subdir_path
+
+    def _download_file(self, download_type, file_url):
+        """
+        _download_file: download execution distributor
+
+        params:
+        download_type: download type for web source file
+        file_url: file URL
+
+        """
+        if download_type == 'Direct Download':
+            copy_file_path = self._download_direct_download_link(file_url)
+        elif download_type == 'DropBox':
+            copy_file_path = self._download_dropbox_link(file_url)
+        elif download_type == 'FTP':
+            copy_file_path = self._download_ftp_link(file_url)
+        elif download_type == 'Google Drive':
+            copy_file_path = self._download_google_drive_link(file_url)
+        else:
+            raise ValueError('Invalid download type: {}'.format(download_type))
+
+        return copy_file_path
+
+    def _retrieve_filepath(self, file_url, download_type):
+        """
+        _retrieve_filepath: retrieve file name from download URL and return local file path
+
+        """
+        try:
+            response = requests.get(file_url, stream=True)
+            content_disposition = response.headers['content-disposition']
+        except KeyError:
+            self.log('Parsing file name directly from URL')
+            file_name = file_url.split('/')[-1]
+        else:
+            file_name = content_disposition.split('filename="')[-1].split('";')[0]
+        finally:
+            response.close()
+            
+        self.log('Retrieving file name from url: {}'.format(file_name))
+        copy_file_path = os.path.join(self.tmp, file_name)
```

Comment:
@MrCreosote 
I see. self.tmp is already scratch/uuid/... Do you think we need another uuid?
`self.tmp = os.path.join(self.scratch, str(uuid.uuid4()))`